### ISSUE TITLE
[Feature Req.] Add X and Y caret policy user configurability

### ISSUE THREAD

**Issue by alankilborn** [notepad-plus-plus/notepad-plus-plus#12020]
It would be a nice feature for Notepad++ users to be able to set the X and Y caret policy provided by Scintilla.

This would provide an edge-of-viewport buffer region, improving data visualization.

As it is currently (N++ 8.4.4), sometimes caret navigation operations (e.g. Ctrl+End) will leave the user with the caret at the righthand window edge.  This does not allow the user to see the text to the right of the caret, without horizontally scrolling (an additional operation).  If a caret policy could be set, the user could adjust this to their general need, and a certain number of characters would be forced to be visible on the right of the viewport after the caret location.

The above is just an example for the right-edge of the editing window.  Similar "problems" currently exist for the left edge, and top and bottom edges as well.

This was first mentioned [HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8961#issuecomment-716670660) and was upvoted by @donho.
It came up again recently in #11979.

The UI for the _Preferences_ > _Editing_ screen needed are:

* `X slop:`, in characters (default: 0)
* `Y slop:`, in lines (default: 0)

Note: The controls for the UI could be sliders with fixed limits (not sure what the limit should be) or, perhaps less desirable, integer input boxes.

Note: using a default of 0 would preserve existing Notepad++ functionality.  Only users that need the improved functionality would need to go into the _Preferences_ and make a change.

The needed Scintilla functions are shown [HERE](https://www.scintilla.org/ScintillaDoc.html#SCI_SETXCARETPOLICY) in the docs.

The suggestion for keeping it simple for the user is to call Scintilla function with the STRICT, EVEN, and SLOP flags.

**Comment by vinsworldcom** [comment]
I love this idea!  I opened the [first issue](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8961) @alankilborn quotes because I was using the feature and it was being stomped on.  Now that it's fixed, making X/Y policy available for end users would be great.

This provides some context when scrolling through large / long lines files by making some "buffer" always visible beyond the cursor scrolling direction.

My simple implementation now is in my `startup.py` for PythonScript and it applies to both Scintilla edit views and then never again, so this wouldn't need to be applied to each new buffer, just set based on the user settings in preferences and then scrolling behavior follows that policy.

Cheers.

**Comment by AndiKiissel** [comment]
In some strange way with this script `ed.setXCaretPolicy(carePolicy, 0)` leaves **1 pixel(!)** slop and `ed.setYCaretPolicy(carePolicy, 0)` **1 line** slop.
NppExec `sci_sendmsg SCI_SETXCARETPOLICY 13 0` and `sci_sendmsg SCI_SETYCARETPOLICY 13 0` leaves the same artifacts.

**Comment by alankilborn** [comment]
>In some strange way with this script ed.setXCaretPolicy(carePolicy, 0) leaves 1 pixel(!) slop and ed.setYCaretPolicy(carePolicy, 0) 1 line slop.

What exactly are you trying to achieve with this?

From the documentation, such a combination looks like it will do exactly what you are saying it does:

![image](https://user-images.githubusercontent.com/77065706/185464710-a86de632-3978-4496-9dd2-ba65202d0ae3.png)

Since the caret can't go to the margin, I suppose a zero `caretSlop` value (which WOULD take it to the margin) would be detected and thus a minimum of 1 set.

**Comment by alankilborn** [comment]
>NppExec sci_sendmsg SCI_SETXCARETPOLICY 13 0 and sci_sendmsg SCI_SETYCARETPOLICY 13 0 leaves the same artifacts.

Note that for playing around with this using NppExec, the first argument uses these bitweights:

* JUMPS : 16
* EVEN: 8
* STRICT: 4
* SLOP: 1

So, 13 sets EVEN=1, STRICT=1, JUMPS=0, SLOP=1.

It is a curious coincidence that reading the table entry I highlighted in yellow above from left to right yields `1101` which is 13 in binary.  Be careful.

**Comment by alankilborn** [comment]
>I did try to restore default situation with vslop=hslop=0, but this did not work as I expected. Perhaps I should do something differently.

Unfortunately there is no "get" calls in Scintilla to retrieve the current values of the X and Y caret policies, only the "set"s.  If there were "get"s, we could use them to see what the "default situation" is for Notepad++.

Currently Notepad++ (8.4.4) never calls the Scintilla functions to set the X and Y caret policies.  Thus, we can infer that the values it is using are the Scintilla defaults.

In the Scintilla code, we find this:

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/08acfe99b8dcfef40c7e7c244fe5c75cab0338b5/scintilla/src/Editor.cxx#L154-L155

>I was thinking, how Preferences > Editing default ... will act and/or how this can be done.

So, I'd think those settings would be used in a new N++ _Preferences_ UI control(s) when the user doesn't want to change X and Y caret policies away from their defaults.

**Comment by AndiKiissel** [comment]
Yes! With these parameters default behaviour will be restored. Thanks! When caret moves to border then with next move text "jumps" by 50 pixels.
But to change these policies to slope>0, `EVEN | STRICT | SLOP` must be used.
Maybe some smarter then me can explain this logic better...

**Comment by RenamedPlayer** [comment]
Can it be that there is no interest in such a useful feature? I just began with Notepad++ and that was the first thing I missed.

**Comment by alankilborn** [comment]
> Can it be that there is no interest in such a useful feature? I just began with Notepad++ and that was the first thing I missed.

Use NppExec or Pythonscript plugins, as shown above, to do it?

**Comment by RenamedPlayer** [comment]
Thank for the advice, I thought that was something special to expert users. `sci_sendmsg SCI_SETYCARETPOLICY 13 5` in NppExec does it.
